modules/mpa/encode.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * @file mpa/encode.c mpa Encode
+ *
+ * Copyright (C) 2016 Symonics GmbH
+ */
+
+#include <re.h>
+#include <baresip.h>
+#include <twolame.h>
+#include <string.h>
+#include <speex/speex_resampler.h>
+#include "mpa.h"
+
+#undef DEBUG 
+
+struct auenc_state {	
+	twolame_options *enc;
+	int channels;
+	SpeexResamplerState *resampler;
+};
+
+
+static void destructor(void *arg)
+{
+	struct auenc_state *aes = arg;
+
+	if (aes->enc)
+		twolame_close(&aes->enc);
+#ifdef DEBUG
+	debug("mpa: encoder destroyed\n");
+#endif
+}
+
+int mpa_encode_update(struct auenc_state **aesp, const struct aucodec *ac,
+		       struct auenc_param *param, const char *fmtp)
+{
+	struct auenc_state *aes;
+	struct mpa_param prm;
+	const struct aucodec *auc = aucodec_find("MPA", 90000, 1);
+	int mpares;
+
+	(void)param;
+
+	if (!aesp || !ac || !ac->ch)
+		return EINVAL;
+
+	aes = *aesp;
+	
+	if (aes) {
+		mem_deref(aes);
+	}
+		
+	aes = mem_zalloc(sizeof(*aes), destructor);
+	aes->enc = twolame_init();
+	if (!aes->enc) {
+		error("mpa: encoder create failed");
+		mem_deref(aes);
+		return ENOMEM;
+	}
+	aes->channels = auc->ch;
+	*aesp = aes;
+#ifdef DEBUG
+	debug("mpa: encoder created %s\n",fmtp);
+#endif
+
+	prm.samplerate = 32000;
+	prm.bitrate    = 128000;
+	prm.layer      = 2;
+	prm.mode       = STEREO;
+	mpa_decode_fmtp(&prm, fmtp);
+
+	mpares = 0;
+#ifdef DEBUG
+	mpares |= twolame_set_verbosity(aes->enc, 5);	
+#else
+	mpares |= twolame_set_verbosity(aes->enc, 0);	
+#endif
+
+	mpares |= twolame_set_mode(aes->enc, prm.mode == SINGLE_CHANNEL ? TWOLAME_MONO : 
+						prm.mode == DUAL_CHANNEL ? TWOLAME_DUAL_CHANNEL : 
+						prm.mode == JOINT_STEREO ? TWOLAME_JOINT_STEREO :
+						prm.mode == STEREO ? TWOLAME_STEREO : TWOLAME_AUTO_MODE);
+	mpares |= twolame_set_version(aes->enc, prm.samplerate < 32000 ? TWOLAME_MPEG2 : TWOLAME_MPEG1);
+	mpares |= twolame_set_bitrate(aes->enc, prm.bitrate/1000);
+	mpares |= twolame_set_in_samplerate(aes->enc, prm.samplerate);
+	mpares |= twolame_set_out_samplerate(aes->enc, prm.samplerate);
+	mpares |= twolame_set_num_channels(aes->enc, 2);
+	if(mpares!=0) {
+		error("mpa: encoder set failed\n");
+		return EINVAL;
```

Comment:
if this fails it will lead to a memory leak of the state object. You should instead
set ``err`` to the error value and then goto out. In the ``out`` label you must handle
both success and error.

(same with other return statements further below).